File: frontend/src/types/wheeling.ts
// Wheeling system types

export interface WheelingGridItem {
  numbers: number[];
  stars: number[] | null;
}

export interface GainScenario {
  rank: number;
  name: string;
  match_numbers: number;
  match_stars: number;
  avg_prize: number;
  matching_grids_best: number;
  matching_grids_avg: number;
  matching_grids_worst: number;
  potential_gain_best: number;
  potential_gain_avg: number;
  potential_gain_worst: number;
}

export interface WheelingPreviewRequest {
  numbers: number[];
  stars?: number[] | null;
  guarantee: number;
}

export interface WheelingPreviewResponse {
  estimated_grid_count: number;
  estimated_cost: number;
  total_t_combinations: number;
  full_wheel_size: number;
  reduction_rate_estimate: number;
}

export interface WheelingGenerateRequest {
  numbers: number[];
  stars?: number[] | null;
  guarantee: number;
}

export interface WheelingGenerateResponse {
  id: number | null;
  grids: WheelingGridItem[];
  grid_count: number;
  total_cost: number;
  coverage_rate: number;
  reduction_rate: number;
  total_t_combinations: number;
  full_wheel_size: number;
  computation_time_ms: number;
  gain_scenarios: GainScenario[];
  number_distribution: Record<number, number>;
}

export interface WheelingSystemResponse {
  id: number;
  game_id: number;
  selected_numbers: number[];
  selected_stars: number[] | null;
  guarantee_level: number;
  grids: WheelingGridItem[];
  grid_count: number;
  total_cost: number;
  coverage_rate: number;
  reduction_rate: number;
  created_at: string;
}

File: backend/app/schemas/wheeling.py
"""Wheeling system Pydantic schemas — request / response models."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


# ── Request schemas ──


class WheelingPreviewRequest(BaseModel):
    numbers: list[int] = Field(..., min_length=6, max_length=20)
    stars: list[int] | None = Field(None, max_length=6)
    guarantee: int = Field(3, ge=2)


class WheelingGenerateRequest(BaseModel):
    numbers: list[int] = Field(..., min_length=6, max_length=20)
    stars: list[int] | None = Field(None, max_length=6)
    guarantee: int = Field(3, ge=2)


# ── Response schemas ──


class WheelingGridItem(BaseModel):
    numbers: list[int]
    stars: list[int] | None = None


class GainScenarioSchema(BaseModel):
    rank: int
    name: str
    match_numbers: int
    match_stars: int
    avg_prize: float
    matching_grids_best: int
    matching_grids_avg: float
    matching_grids_worst: int
    potential_gain_best: float
    potential_gain_avg: float
    potential_gain_worst: float


class WheelingPreviewResponse(BaseModel):
    estimated_grid_count: int
    estimated_cost: float
    total_t_combinations: int
    full_wheel_size: int
    reduction_rate_estimate: float


class WheelingGenerateResponse(BaseModel):
    id: int | None = None
    grids: list[WheelingGridItem]
    grid_count: int
    total_cost: float
    coverage_rate: float
    reduction_rate: float
    total_t_combinations: int
    full_wheel_size: int
    computation_time_ms: float
    gain_scenarios: list[GainScenarioSchema] = []
    number_distribution: dict[int, int] = {}


class WheelingSystemResponse(BaseModel):
    id: int
    game_id: int
    selected_numbers: list[int]
    selected_stars: list[int] | None
    guarantee_level: int
    grids: list[WheelingGridItem]
    grid_count: int
    total_cost: float
    coverage_rate: float
    reduction_rate: float
    created_at: datetime

    model_config = {"from_attributes": True}
